Ajacobs/pagerduty plugin service id annotation

alecjacobs5401 opened the pull request:
Resolves #11667

## Hey, I just made a Pull Request!

Introduces a new annotation `pagerduty.com/service-id` that can be used instead of the `pagerduty.com/integration-key` annotation.
_Note: If both annotations are specified on a given Entity, then the `pagerduty.com/integration-key` annotation will be prefered_

In addition, various enhancements/bug fixes were introduced:

- The `PagerDutyCard` component now wraps error and loading messages with an `InfoCard` to contain errors/messages. This enforces a consistent experience on the EntityPage
- If no service can be found for the provided integration key, a new Error Message Empty State component will be shown instead of an error alert
- Introduces the `fetchApi` to replace standard `window.fetch`
  - ensures that Identity Authorization is respected and provided in API requests

## Screen Shots
### Missing API Token
**Before**
![image](https://user-images.githubusercontent.com/12898497/170779745-d01e10e4-8ea5-451c-9840-6c9bb01a5154.png)

**After**
![image](https://user-images.githubusercontent.com/12898497/170778858-9a2d208b-cc82-47e6-94b0-ee4c73ec0e90.png)

### Entity Not Found
**Before**
![image](https://user-images.githubusercontent.com/12898497/170780023-1f6e63ed-e70f-45f7-a517-f7f8f147c55d.png)

**After**
![image](https://user-images.githubusercontent.com/12898497/170779143-003b703e-2380-4649-acb3-bb64e690c17c.png)

### Component Annotated with `pagerduty.com/service-id`
![image](https://user-images.githubusercontent.com/12898497/170779280-1704d449-663e-481f-bf11-a48db6c327dd.png)

### :heavy_check_mark: Checklist

<!--- Please include the following in your Pull Request when applicable: -->

- [x] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [x] Added or updated documentation
- [x] Tests for new functionality and regression tests for bug fixes
- [x] Screenshots attached (for UI changes)
- [x] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

alecjacobs5401 commented on plugins/pagerduty/src/components/PagerDutyCard/index.tsx:
```diff
@@ -67,33 +76,54 @@ export const PagerDutyCard = () => {
+        );
+    }
+
+    return <BasicCard>{errorNode}</BasicCard>;
```
wrapping the empty state errors in this `BasicCard` _does_ unfortunately have an undesirable "dark background over light background over dark background" effect when in dark mode. Happy to remove this for errors, but the consistent Info Card seemed desirable

benjdlambert reviewed (commented):
Looking pretty good, just a comment about the errors.

alecjacobs5401 commented on plugins/pagerduty/src/api/client.ts:
```diff
@@ -29,28 +30,38 @@ import {
 } from '@backstage/core-plugin-api';
 
 export class UnauthorizedError extends Error {}
+export class NotFoundError extends Error {}
```
I thought about that but I wasn't sure if those were more specific to catalog for some reason (even though there is a dedicated generic package)

I'll definitely go ahead and update to use that

alecjacobs5401 commented:
@benjdlambert Thanks for the review! Adjusted to use `@backstage/errors` based on feedback and ensured that `yarn lint` is successful

benjdlambert commented on .changeset/weak-bananas-deliver.md:
```diff
@@ -0,0 +1,13 @@
+---
+'@backstage/plugin-pagerduty': patch
```
This should be a minor change, as we're technically breaking the `fromConfig` method in the `api-report.md`

alecjacobs5401 replied:
Makes sense! I was conflicted about this initially since I would expect the client is setup automatically and wouldn't require manual intervention, but thats likely an incorrect generalization.

Updated to `minor`!

alecjacobs5401 commented:
@benjdlambert Appreciate your reviews here!

Pushed up the fix to make this a `minor` change bump over `patch`. Also rebased at your suggestion.

Let me know if anything else jumps out at you!

Rugvip reviewed (commented):
Thank you! 👍 

Sorry for joining in late, but spotted a few things in the API surface that I think we could iterate on a bit.

alecjacobs5401 commented on plugins/pagerduty/api-report.md:
```diff
@@ -64,6 +64,8 @@ export class PagerDutyClient implements PagerDutyApi {
   // (undocumented)
   getServiceByIntegrationKey(integrationKey: string): Promise<Service[]>;
+  // (undocumented)
+  getServiceByServiceId(serviceId: string): Promise<Service>;
```
@Rugvip thanks for looking!

>On a separate note, return type here is a bit suspicious, the old one returns an array, but this one is just a single service?

Yeah, the original implementation queries the [List Services Endpoint](https://developer.pagerduty.com/api-reference/e960cca205c0f-list-services) with a filter on the integration key, so the return value of that API is an array of services.

This new implementation is specific in querying an exact service using the [Get Service Endpoint](https://developer.pagerduty.com/api-reference/165ad96a22ffd-get-a-service), so there is only one object to return.

>I feel it would be better to move this all the way to getServiceByEntity()

Noted! I didn't feel comfortable making that decision on my own, so I just went ahead an bolted on an option to better support existing behavior.

I'll go ahead and refactor a bit, but I think that its currently configured this way to isolate dependencies on React and the API Client.

> essentially baking the logic from `usePagerdutyEntity` into the API implementation rather than having that live in React land

This may not be _totally_ feasible since [`usePagerdutyEntity` relies on useEntity from the `@backstage/plugin-catalog-react`](https://github.com/backstage/backstage/blob/32bff53079c7b6890067f515be06e291804faed2/plugins/pagerduty/src/hooks/index.ts#L17), so there is an implied reliance on doing some of this login _in react_. 

What you're suggesting feels like a whole refactor of the API client, since there are other API calls that rely on on providing a `serviceID` to query data (e.g. [`getIncidentsByServiceId`](https://github.com/backstage/backstage/blob/79b0e20cb2bb62963e223b884dd6a542f2119f42/plugins/pagerduty/src/api/client.ts#L83) and [`getChangeEventsBySeviceId`](https://github.com/backstage/backstage/blob/79b0e20cb2bb62963e223b884dd6a542f2119f42/plugins/pagerduty/src/api/client.ts#L93))

I'm inclined to take what you're suggesting, but moreso have the `getServiceByEntity` take an object of type `PagerDutyEntity` that we can normalize.

I'll take a stab at putting something together!

alecjacobs5401 commented:
@Rugvip @benjdlambert @jhaals 

Thanks for the reviews everyone!

The last round of changes, I may have gone overboard to make the API Client consistent with the feedback provided by @Rugvip [here](https://github.com/backstage/backstage/pull/11697#discussion_r888977534)

I believe that I addressed all of the comments there, I know it isn't _exactly_ the same as what was suggested, but I think its a good next step. 

Open to all other feedback! Please let me know if I can/should change anything else.

alecjacobs5401 commented:
@Rugvip @benjdlambert @jhaals 

Bumping another review of this PR - thanks in advance!

alecjacobs5401 commented:
Thanks for letting the workflows run!

I took a look at the failing [E2E Text for Linux](https://github.com/backstage/backstage/runs/6901841536?check_suite_focus=true) and it looks like it failed during
```
yarn tsc:full
```
but I wasn't able to reproduce the failure locally on my end (probably because I'm not running linux).

Any advice there?

Rugvip commented:
@alecjacobs5401 it's now fixed so a rebase/merge should do the trick!

alecjacobs5401 commented:
@Rugvip rebased as requested. Thanks!

Rugvip commented on plugins/pagerduty/api-report.md:
```diff
@@ -10,21 +10,113 @@ import { BackstagePlugin } from '@backstage/core-plugin-api';
+// Warning: (ae-missing-release-tag) "ClientApiConfig" is exported by the package, but it is missing a release tag (@alpha, @beta, @public, or @internal)
+//
+// @public (undocumented)
+export type ClientApiConfig = ClientApiDependencies & {
```
```suggestion
export type PagerDutyClientConfig = PagerDutyClientDependencies & {
```

Rugvip commented on plugins/pagerduty/api-report.md:
```diff
@@ -87,4 +210,15 @@ export function TriggerButton(props: TriggerButtonProps): JSX.Element;
+// Warning: (ae-missing-release-tag) "User" is exported by the package, but it is missing a release tag (@alpha, @beta, @public, or @internal)
+//
+// @public (undocumented)
+export type User = {
```
These type names end up being extremely generic, which I think can hurt projects using this package down the line. Imagine 50 plugins similar to this installed in a project where you want to autocomplete an import, kinda sucks to have to figure out which of the 30 different `User` types that you actually are supposed to import.

What do you think by prefixing pretty much all these added types by `PagerDuty*`? `PagerDutyUser`, `PagerDutyServiceResponse`, etc.

Rugvip commented on plugins/pagerduty/api-report.md:
```diff
@@ -40,43 +132,74 @@ export const PagerDutyCard: () => JSX.Element;
-  getServiceByIntegrationKey(integrationKey: string): Promise<Service[]>;
-  // Warning: (ae-forgotten-export) The symbol "TriggerAlarmRequest" needs to be exported by the entry point index.d.ts
-  //
+  getServiceByEntity(
+    pagerDutyEntity: PagerDutyEntity,
+  ): Promise<ServiceResponse>;
```
Note quite what was intended here 😅, just pass the regular `Entity` straight from `useEntity` into this method, then do any unpacking of annotations and so on in the implementation.

Rugvip replied:
Btw, I realized there are other lookup and haven't investigated in-depth, but my assumption here is that we can use `getServiceByEntity` to look up the initial service, and then use the ID of that service in all the other places?

Rugvip reviewed (commented):
Another round, think we'll be good to go after these 👍

alecjacobs5401 commented:
@Rugvip I believe I have addressed your comments.

Thanks in advance!

alecjacobs5401 commented:
Awesome, thanks!
